src/module/actor/helpers.ts

```diff
@@ -252,6 +253,62 @@ function createEncounterRollOptions(actor: ActorPF2e): Record<string, boolean> {
     return Object.fromEntries(entries);
 }
 
+/** Create roll options pertaining to the terrain the actor is currently in */
+function createTerrainRollOptions(actor: ActorPF2e): Record<string, boolean> {
+    const toAdd = new Set<string>();
+    // Always add the scene terrain types
+    for (const terrain of canvas.scene?.flags.pf2e.terrainTypes ?? []) {
+        toAdd.add(terrain);
+    }
+    const token = actor.getActiveTokens(false, true).at(0);
+    const terrains = ((): Set<string> => {
+        // No token on the scene means no terrain roll options
+        if (!token) return new Set<string>();
+        const toRemove = new Set<string>();
+        for (const region of token.regions ?? []) {
+            // An elevation value of null translates to Infinity
+            const bottom = region.elevation.bottom ?? -Infinity;
+            const top = region.elevation.top ?? Infinity;
+            if (token.elevation < bottom || token.elevation > top) continue;
+
+            for (const behavior of region.behaviors.filter((b) => b.type === "pf2eTerrain")) {
+                // todo: remove once type resolution is possible
+                const system = behavior.system as TerrainTypeData;
+                switch (system.mode) {
+                    case "add": {
+                        for (const terrain of system.terrainTypes) {
+                            toAdd.add(terrain);
+                        }
+                        break;
+                    }
+                    case "remove": {
+                        for (const terrain of system.terrainTypes) {
+                            toRemove.add(terrain);
+                        }
+                        break;
+                    }
+                    case "override": {
+                        // Only clear out the exisiting values in case there is another
+                        // behavior after the override
+                        toAdd.clear();
+                        toRemove.clear();
```

Comment:
Don't you need to remove the base terrains?

Reply:
The scene terrains are added to the `toAdd` set at the start so they are cleared out here too.

Reply:
oh! I see.